### PR TITLE
sd-dhcp-server: remove lease with hashmap_remove_value()

### LINKED ISSUE #22253: [systemd-networkd] regular crashes since v250 release
**systemd version the issue has been seen with**

```
systemd 250 (250.3-1-arch)
+PAM +AUDIT -SELINUX -APPARMOR -IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT +QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +BPF_FRAMEWORK +XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified
```
**Used distribution**

Latest Arch Linux.

**Linux kernel version used** (`uname -a`)

```
$ uname -a
Linux defiant 5.16.0-pf2 #1 SMP PREEMPT Sun, 23 Jan 2022 08:47:42 +0000 x86_64 GNU/Linux
```

(_it was the same with v5.15_)

**CPU architecture issue was seen on**

`x86_64`

**Expected behaviour you didn't see**

`systemd-neworkd` should not segfault.

**Unexpected behaviour you saw**

`systemd-neworkd` segfaults once DHCP client connects via Wi-Fi.

**Steps to reproduce the problem**

There's a static lease defined for device:

```
[DHCPServerStaticLease]
MACAddress=xx:xx:xx:xx:xx:xx
Address=1.2.3.4
```

The Wi-Fi interface is used by hostapd, but hostapd doesn't do anything with bridging, VLANs etc. All the networking stuff is on `systemd-networkd`. The Wi-Fi interface is a part of bridge:

```
[Match]
Name=wlp1s0

[Network]
Bridge=wifi-br
```

The bridge is defined as following:

```
[NetDev]
Name=wifi-br
Kind=bridge
```

```
[Match]
Name=wifi-br

[Network]
LinkLocalAddressing=no
IPv6AcceptRA=no
```

The bridge also contains a gateway interface:

```
[NetDev]
Name=wifi-gw
Kind=veth

[Peer]
Name=wifi-gw-br
```

```
[Match]
Name=wifi-gw-br

[Network]
Bridge=wifi-br
LinkLocalAddressing=no
IPv6AcceptRA=no
```

```
[Match]
Name=wifi-gw

[Network]
Address=1.2.3.1/24
LinkLocalAddressing=no
IPv6AcceptRA=no
DHCPServer=yes

[DHCPServer]
DNS=1.2.3.1
NTP=1.2.3.1
SendOption=15:string:example.home.arpa
```

**Additional program output to the terminal or log subsystem illustrating the issue**

Once `systemd-networkd` crashes the following stack trace applies:

```
Core was generated by `/usr/lib/systemd/systemd-networkd'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f122d9dfb30 in main_arena () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007f122d9dfb30 in main_arena () from /usr/lib/libc.so.6
#1  0x00007f122dba2c78 in base_bucket_hash (h=h@entry=0x7f122d9dfb50 <main_arena+336>, p=p@entry=0x55c9d5366648) at ../systemd-stable/src/basic/hashmap.c:341
#2  0x00007f122dba574f in _hashmap_get (h=0x7f122d9dfb50 <main_arena+336>, key=key@entry=0x55c9d5366648) at ../systemd-stable/src/basic/hashmap.c:1350
#3  0x000055c9d3845493 in hashmap_get (key=0x55c9d5366648, h=<optimized out>) at ../systemd-stable/src/basic/hashmap.h:176
#4  dhcp_lease_free (lease=lease@entry=0x55c9d5366640) at ../systemd-stable/src/libsystemd-network/sd-dhcp-server.c:34
#5  0x000055c9d3846e56 in dhcp_server_cleanup_expired_leases (server=0x55c9d532a310) at ../systemd-stable/src/libsystemd-network/sd-dhcp-server.c:874
#6  dhcp_server_handle_message (length=<optimized out>, message=0x55c9d53429b0, server=0x55c9d532a310) at ../systemd-stable/src/libsystemd-network/sd-dhcp-server.c:921
#7  server_receive_message (s=<optimized out>, fd=<optimized out>, revents=<optimized out>, userdata=<optimized out>)
at ../systemd-stable/src/libsystemd-network/sd-dhcp-server.c:1245
#8  0x00007f122dc70d9e in source_dispatch (s=0x55c9d535f420) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:3598
#9  0x00007f122dc70f8e in sd_event_dispatch (e=0x55c9d52fd1c0) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:4183
#10 0x00007f122dc727a9 in sd_event_run (e=0x55c9d52fd1c0, timeout=18446744073709551615) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:4244
#11 0x00007f122dc72980 in sd_event_loop (e=0x55c9d52fd1c0) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:4265
#12 0x000055c9d379eb2b in run (argv=<optimized out>, argc=<optimized out>) at ../systemd-stable/src/network/networkd.c:114
#13 main (argc=<optimized out>, argv=<optimized out>) at ../systemd-stable/src/network/networkd.c:121
```

Sometimes it is different:

```
free(): double free detected in tcache 2
```

```
Core was generated by `/usr/lib/systemd/systemd-networkd'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00007f0cb42a3d22 in raise () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007f0cb42a3d22 in raise () from /usr/lib/libc.so.6
#1  0x00007f0cb428d862 in abort () from /usr/lib/libc.so.6
#2  0x00007f0cb42e5d28 in __libc_message () from /usr/lib/libc.so.6
#3  0x00007f0cb42ed92a in malloc_printerr () from /usr/lib/libc.so.6
#4  0x00007f0cb42ef408 in _int_free () from /usr/lib/libc.so.6
#5  0x00007f0cb42f29e8 in free () from /usr/lib/libc.so.6
#6  0x000055e7575104bb in dhcp_lease_free (lease=lease@entry=0x55e757850280) at ../systemd-stable/src/libsystemd-network/sd-dhcp-server.c:47
#7  0x000055e757511e56 in dhcp_server_cleanup_expired_leases (server=0x55e75783b700) at ../systemd-stable/src/libsystemd-network/sd-dhcp-server.c:874
#8  dhcp_server_handle_message (length=<optimized out>, message=0x55e757866f80, server=0x55e75783b700) at ../systemd-stable/src/libsystemd-network/sd-dhcp-server.c:921
#9  server_receive_message (s=<optimized out>, fd=<optimized out>, revents=<optimized out>, userdata=<optimized out>)
at ../systemd-stable/src/libsystemd-network/sd-dhcp-server.c:1245
#10 0x00007f0cb46b8d9e in source_dispatch (s=0x55e757836200) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:3598
#11 0x00007f0cb46b8f8e in sd_event_dispatch (e=0x55e75780e1c0) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:4183
#12 0x00007f0cb46ba7a9 in sd_event_run (e=0x55e75780e1c0, timeout=18446744073709551615) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:4244
#13 0x00007f0cb46ba980 in sd_event_loop (e=0x55e75780e1c0) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:4265
#14 0x000055e757469b2b in run (argv=<optimized out>, argc=<optimized out>) at ../systemd-stable/src/network/networkd.c:114
#15 main (argc=<optimized out>, argv=<optimized out>) at ../systemd-stable/src/network/networkd.c:121
```

The issue doesn't happen in 100% of cases, but pretty often (_≈75% maybe_).

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #22253.

### CODE REVIEW

**Review by bluca — APPROVED** [review]

**Comment by yuwata** [comment]
@bluca I added one more commit. Could you review again?

**Comment by yuwata** [comment]
Note, the issue is caused by b713a99b1a04507412b387214a26e43a2916bd01 (v250). Please backport this to v250-stable.

**Review by bluca — APPROVED** [review]

**Comment by yuwata** [comment]
@bluca Thank you!

**Comment by bluca** [comment]
Could this be related?

```
17:36:40 ======================================================================
17:36:40 FAIL: test_activation_policy (__main__.NetworkdNetworkTests) (iteration=1, expect_up=True, test='down')
17:36:40 ----------------------------------------------------------------------
17:36:40 Traceback (most recent call last):
17:36:40   File "/build/./test/test-network/systemd-networkd-tests.py", line 3076, in _test_activation_policy
17:36:40     self.wait_operstate('test1', operstate, setup_state=setup_state, setup_timeout=20)
17:36:40   File "/build/./test/test-network/systemd-networkd-tests.py", line 625, in wait_operstate
17:36:40     self.fail(f'Timed out waiting for {link} to reach state {operstate}/{setup_state}')
17:36:40 AssertionError: Timed out waiting for test1 to reach state routable/configured
```

**Comment by evverx** [comment]
@yuwata could you add a test? I'm not sure that particular scenario can be easily reproduced with the integration test but I think it should be possible to do that in `test-dhcp-server`.

**Comment by evverx** [comment]
Looking at https://coveralls.io/builds/45954594/source?filename=src%2Flibsystemd-network%2Fsd-dhcp-server.c#L903 it seems it isn't covered by any tests. @mrc0mmand just out of curiosity I wonder if `systemd-networkd-tests.py` is included in that coverage report? Given that it's suspiciously green I suspect it is but just to be sure :-)

**Comment by yuwata** [comment]
> Could this be related?
> 
> ```
> 17:36:40 ======================================================================
> 17:36:40 FAIL: test_activation_policy (__main__.NetworkdNetworkTests) (iteration=1, expect_up=True, test='down')
> 17:36:40 ----------------------------------------------------------------------
> 17:36:40 Traceback (most recent call last):
> 17:36:40   File "/build/./test/test-network/systemd-networkd-tests.py", line 3076, in _test_activation_policy
> 17:36:40     self.wait_operstate('test1', operstate, setup_state=setup_state, setup_timeout=20)
> 17:36:40   File "/build/./test/test-network/systemd-networkd-tests.py", line 625, in wait_operstate
> 17:36:40     self.fail(f'Timed out waiting for {link} to reach state {operstate}/{setup_state}')
> 17:36:40 AssertionError: Timed out waiting for test1 to reach state routable/configured
> ```

Filed at #22267.

**Comment by yuwata** [comment]
Updated.
- Previously separated two commits is merged, as both commits are necessary to fix the issue.
- Added tests for static lease.
- Added several cleanups.

PTAL.

**Comment by mrc0mmand** [comment]
> Looking at https://coveralls.io/builds/45954594/source?filename=src%2Flibsystemd-network%2Fsd-dhcp-server.c#L903 it seems it isn't covered by any tests. @mrc0mmand just out of curiosity I wonder if `systemd-networkd-tests.py` is included in that coverage report? Given that it's suspiciously green I suspect it is but just to be sure :-)

It should be:
 * https://github.com/systemd/systemd-centos-ci/blob/750c38923d4046f4bf98fe86f9ccf1f21b700d42/vagrant/test_scripts/test-arch-coverage.sh#L193
 * https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-cron-build/lastSuccessfulBuild/artifact/systemd-centos-ci/artifacts_all/artifacts_w3hojuz2/vagrant-logs.ZLQ/vagrant-arch-coverage-testsuite.STO/lcov_networkd_collect_coverage_PASS.log

Hopefully it's doing its job :-)

**Comment by evverx** [comment]
@yuwata thanks! I'm not sure the test can trigger the crash with the first two commits reverted. I wonder if the idea was to cover something else?

> Hopefully it's doing its job :-)

@mrc0mmand it's going strong as far as I can see :-)

**Comment by yuwata** [comment]
I noticed that the first commit just hide the issue.

**Comment by yuwata** [comment]
The previous version just hides the issue, now the root of the issue is also fixed. PTAL!

**Comment by yuwata** [comment]
This becomes huge than I expected, but most commits are trivial cleanups.

**Comment by evverx** [comment]
@yuwata as far as I can see, the fuzzer kind of covers `lease->server->bound_leases_by_*` but `static_leases_by_*` are always empty there. Could you add them to the fuzzer to let it cover that part as well? It would be even better if they could expire there.

**Comment by yuwata** [comment]
@evverx Done. PTAL.

**Comment by evverx** [comment]
@yuwata thanks! CIFuzz is kind of broken now so it didn't actually run the fuzzer but I built it manually and ran for about 5 minutes under ASan/UBSan/MSan and it seems to be working. The PR looks good to me but since the PR is different from what was tested in https://github.com/systemd/systemd/issues/22253 I'd probably wait for @pfactum to confirm that the issue is gone (and new issues aren't introduced :-)).

**Comment by yuwata** [comment]
Note without the above fixes for sd-dhcp-server, the test https://github.com/systemd/systemd/pull/22254/commits/5e61672762701553ca4370d04b2020c2ec7cc88f with small change below can reproduce the original issue.
```diff
diff --git a/src/libsystemd-network/test-dhcp-server.c b/src/libsystemd-network/test-dhcp-server.c
index c3d18ad825..912a8ddee6 100644
--- a/src/libsystemd-network/test-dhcp-server.c
+++ b/src/libsystemd-network/test-dhcp-server.c
@@ -190,22 +190,22 @@ static void test_message_handler(void) {
         /* request address reserved for static lease (unmatching client ID) */
         test.option_client_id.id[6] = 'H';
         test.option_requested_ip.address = htobe32(INADDR_LOOPBACK + 42);
-        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) == 0);
+        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) == DHCP_ACK);
 
         /* request unmatching address */
         test.option_client_id.id[6] = 'G';
         test.option_requested_ip.address = htobe32(INADDR_LOOPBACK + 41);
-        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) == 0);
+        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) == DHCP_ACK);
 
         /* request matching address */
         test.option_client_id.id[6] = 'G';
         test.option_requested_ip.address = htobe32(INADDR_LOOPBACK + 42);
-        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) == DHCP_ACK);
+        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) < 0);
 
         /* try again */
         test.option_client_id.id[6] = 'G';
         test.option_requested_ip.address = htobe32(INADDR_LOOPBACK + 42);
-        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) == DHCP_ACK);
+        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) < 0);
 }
 
 static uint64_t client_id_hash_helper(DHCPClientId *id, uint8_t key[HASH_KEY_SIZE]) {
```

**Comment by pfactum** [comment]
Do I apply all 19 commits for my test, or there's some minimal subset that fixes the issue I've reported?

**Comment by yuwata** [comment]
> Do I apply all 19 commits for my test, or there's some minimal subset that fixes the issue I've reported?

The commits 8e933ee81b01c95b5c596a4e7229235795c6f0da..43227d5c553a169febf11e323c994bba445350a9 are necessary.
But there exists one conflict when these are applied to v250. I prepared a version which can be applied to v250.
Please try https://github.com/systemd/systemd-stable/pull/160.

**Comment by yuwata** [comment]
@pfactum If you also want to backport the reproducer of the issue, then try 5e61672.
With the commits in https://github.com/systemd/systemd-stable/pull/160, the test should pass, but the without the fixes, the test should fail. See https://github.com/systemd/systemd/pull/22254#issuecomment-1023857704 for more detail.

**Comment by pfactum** [comment]
With https://github.com/systemd/systemd-stable/pull/160 the crash is not reproducible for me.

**Comment by yuwata** [comment]
@pfactum Thank you for your help!

File: src/libsystemd-network/sd-dhcp-server.c [hunk 1]
```diff
@@ -1133,14 +1133,17 @@ int dhcp_server_handle_message(sd_dhcp_server *server, DHCPMessage *message, siz
                         return DHCP_ACK;
                 }
 
-                DHCPLease *existing_lease_by_address = hashmap_get(server->bound_leases_by_address, UINT32_TO_PTR(address));
-                if (address_is_in_pool(server, address) && existing_lease_by_address == existing_lease) {
+                if (address_is_in_pool(server, address)) {
+                        /* The requested address is in the pool. */
+
+                        if (existing_lease && existing_lease->address != address)
+                                /* We previously assign an address, but the client requests another address. Refusing. */
```
Comment (bluca) [inline src/libsystemd-network/sd-dhcp-server.c]: "We previously assigned an address, but the client rested another one. Refuse."

File: src/libsystemd-network/sd-dhcp-server.c [hunk 2]
```diff
@@ -1084,9 +1084,13 @@ int dhcp_server_handle_message(sd_dhcp_server *server, DHCPMessage *message, siz
...
                         return 0;
 
-                /* verify that the requested address is from the pool, and either
-                   owned by the current client or free */
-                if (static_lease && static_lease->address == address) {
+                if (static_lease) {
+                        /* Found a static lease for the client ID. */
+
+                        if (static_lease->address != address)
+                                /* The client requests an address which is different from the static lease. Refusing. */
```
Comment (bluca) [inline src/libsystemd-network/sd-dhcp-server.c]: "The client requested an address..."
"Refuse."

File: src/libsystemd-network/sd-dhcp-server.c [hunk 3]
```diff
@@ -612,16 +612,24 @@ static int server_send_offer_or_ack(
         return dhcp_server_send_packet(server, req, packet, type, offset);
 }
 
-static int server_send_nak(sd_dhcp_server *server, DHCPRequest *req) {
+static int server_send_nak_or_ignore(sd_dhcp_server *server, bool init_reboot, DHCPRequest *req) {
         _cleanup_free_ DHCPPacket *packet = NULL;
         size_t offset;
         int r;
 
+        if (!init_reboot)
```
Comment (bluca) [inline src/libsystemd-network/sd-dhcp-server.c:627]: Feel free to ignore, but to me it seemed better with this condition checked outside of the function - it doesn't seem very related. But not too important.

**Review by bluca — APPROVED** [review]

**Review by anitazha — APPROVED** [review]
LGTM

File: src/libsystemd-network/sd-dhcp-server.c (hunk 3 above)
Comment (yuwata) [reply src/libsystemd-network/sd-dhcp-server.c:627]: I'd like to keep this as is. This makes `dhcp_server_handle_message()` simpler.

Instead, I added a brief comment why we silently ignore the request in that case.

**Comment by yuwata** [comment]
@bluca Thank you. The comments are updated. Upgrading the green label.

**Pull request merged**